### PR TITLE
refactor: unify format tool creation logic

### PR DESCRIPTION
**Pull request by teamdandelion**
Now the formatting module creates a tool schema, and the providers use
their existing schema -> provider specific tool represetnation pipeline.

I modified the standard structured output test case to include nested
models and this had the salubrious effect of testing tools with nested
models as function arguments; particularly I had to fix a bug with
resolving refs for Google usage.

### CODE REVIEW

**Comment by teamdandelion** [comment]
* **#1385** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1385?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#1386](https://github.com/Mirascope/mirascope/pull/1386) <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1386?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1388](https://github.com/Mirascope/mirascope/pull/1388) <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1388?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1384** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1384?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1383** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1383?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1382** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1382?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1381** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1381?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1380** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1380?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1379** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1379?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1378** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1378?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1377** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1377?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1376** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1376?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1375** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1375?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1374** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1374?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1373** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1373?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1372** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1372?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1371** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1371?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1370** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1370?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1369** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1369?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1368** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1368?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1366** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1366?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1367** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1367?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1365** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1365?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1364** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1364?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1363** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1363?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1362** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1362?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1361** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1361?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1360** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1360?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1359** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1359?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1358** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1358?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1357** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1357?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1356** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1356?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1355** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1355?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1354** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1354?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1352** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1352?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1351** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1351?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1353** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1353?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1353?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1346** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1346?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1350** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1350?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1349** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1349?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1348** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1348?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1347** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1347?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1345** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1345?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1344** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1344?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1343** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1343?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1342** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1342?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1341** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1341?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1340** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1340?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1339** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1339?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1338** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1338?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1337** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1337?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1335** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1335?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1336** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1336?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1334** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1334?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1333** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1333?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1330** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1330?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1331](https://github.com/Mirascope/mirascope/pull/1331) <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1331?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1332** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1332?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1329** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1328** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1328?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1327** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1327?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1326** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1326?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1325** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1325?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1324** <a href="https://app.graphite.dev/github/pr/Mirascope/mirascope/1324?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `v2`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python/mirascope/llm/clients/google/_utils.py [hunk 1]
```diff
@@ -54,6 +59,26 @@
...
+) -> dict[str, Any]:
+    """Recursively resolve $ref references in a JSON Schema object."""
+    if isinstance(schema, dict):
+        if "$ref" in schema:
+            ref_path = schema["$ref"]
+            if ref_path.startswith("#/$defs/"):
+                ref_name = ref_path.split("/")[-1]
+                if defs and ref_name in defs:
+                    return _resolve_refs(defs[ref_name], defs)
+            return schema  # pragma: no cover
```
Comment (teamdandelion) [inline python/mirascope/llm/clients/google/_utils.py:73]: Set no cover on this edge case wherein we have a ref that doesn't start with `"#/$defs/"` or isn't in defs. I no-covered it rather than write unit tests for this since it's a private utility method and I'm content with getting coverage from our real use cases.

File: python/mirascope/llm/clients/openai/_utils.py [hunk 2]
```diff
@@ -222,6 +228,34 @@ def _convert_tool_to_tool_param(
     )
 
 
+def _create_strict_response_format(
```
Comment (teamdandelion) [inline python/mirascope/llm/clients/openai/_utils.py:231]: Moving this up in the file per request downstack

File: python/tests/e2e/cassettes/structured_output/anthropic/async.yaml [hunk 3]
```diff
@@ -5,9 +5,10 @@ interactions:
       respond to the user''s query using the __mirascope_formatted_output_tool__ tool
       for structured output.","tool_choice":{"type":"tool","name":"__mirascope_formatted_output_tool__","disable_parallel_tool_use":true},"tools":[{"name":"__mirascope_formatted_output_tool__","description":"Use
       this tool to extract data in Book format for a final response.\nA book with
-      a rating. The title should be in all caps!","input_schema":{"description":"A
-      book with a rating. The title should be in all caps!","properties":{"title":{"title":"Title","type":"string"},"author":{"title":"Author","type":"string"},"rating":{"description":"For
```
Comment (teamdandelion) [inline python/tests/e2e/cassettes/structured_output/anthropic/async.yaml:9]: nb (here and elsewhere): This fixes an issue I've noted in the past, where we would include the description twice. ("The title should be in all caps!" is present twice in the left and only once on the right.)

File: python/mirascope/llm/clients/google/_utils.py [hunk 4]
```diff
@@ -165,14 +190,21 @@ def _convert_tool_to_function_declaration(
     """Convert a single Mirascope tool to Google FunctionDeclaration format with caching."""
     schema_dict = tool.parameters.model_dump(by_alias=True, exclude_none=True)
     schema_dict["type"] = "object"
+
+    defs = schema_dict.get("$defs")
+
```
Comment (willbakst) [inline python/mirascope/llm/clients/google/_utils.py]: drop this newline

File: python/mirascope/llm/formatting/_utils.py [hunk 5]
```diff
@@ -25,3 +25,50 @@ def default_formatting_instructions(
         json_schema = json.dumps(schema, indent=2)
         instructions = JSON_MODE_INSTRUCTIONS.format(json_schema=json_schema)
         return inspect.cleandoc(instructions)
+
+
+def to_tool_schema(format: Format[FormattableT]) -> ToolSchema:
```
Comment (willbakst) [inline python/mirascope/llm/formatting/_utils.py]: nit: rename to `create_tool_schema`?

File: python/tests/e2e/test_structured_output.py [hunk 6]
```diff
@@ -25,11 +25,17 @@ def test_structured_output_sync(
 ) -> None:
     """Test synchronous structured output without context."""
 
+    class Author(BaseModel):
```
Comment (willbakst) [inline python/tests/e2e/test_structured_output.py]: I think we can safely define `Author` and `Book` just once at the top of the file?

Comment (teamdandelion) [reply python/tests/e2e/test_structured_output.py]: Yes! Good point. That wasn't safe prior to the refactor to the format system (because the decorator mutated the class), however now that we no longer mutate the classes, this is fine. Done. I updated the other tests in this PR to use the same pattern while here.

Comment (willbakst) [reply python/tests/e2e/test_structured_output.py]: I'm not seeing this change?

**Review by willbakst — APPROVED** [review]

**Comment by teamdandelion** [comment]
### Merge activity

* **Sep 25, 5:33 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1353).
* **Sep 25, 5:35 PM UTC**: [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1353) rebased this pull request as part of a merge.
* **Sep 25, 5:36 PM UTC**: @teamdandelion merged this pull request with [Graphite](https://app.graphite.dev/github/pr/Mirascope/mirascope/1353).
